### PR TITLE
Fix #1523, add --json-errors flag.

### LINKED ISSUE #1523: Support JSON errors

### PR DESCRIPTION
**Pull request by paf31**
@kRITZCREEK @nwolverson @garyb How does this look?

Thanks.

### CODE REVIEW

**Comment by kRITZCREEK** [comment]
Looks good from skimming the changes. I can't test it before this weekend though. One thing that seems strange is to have the types for errors inside the Main module. Other than that super happy to get JSON errors ;) :+1:

File: psc/Main.hs [hunk 1]
```diff
@@ -46,39 +51,90 @@ data PSCMakeOptions = PSCMakeOptions
...
+data JSONResult = JSONResult
+  { warnings :: [JSONError]
+  , errors :: [JSONError]
+  }
+
+$(A.deriveJSON A.defaultOptions ''ErrorPosition)
+$(A.deriveJSON A.defaultOptions ''JSONError)
+$(A.deriveJSON A.defaultOptions ''JSONResult)
+
+-- | Argumnets: verbose, use JSON, warnings, errors
```
Comment (garyb) [inline psc/Main.hs]: Typo :smile:

File: src/Language/PureScript/Errors.hs [hunk 2]
```diff
@@ -184,6 +185,23 @@ data HintCategory
...
+-- | Get the source span for an error
+errorSpan :: ErrorMessage -> SourceSpan
+errorSpan = fromMaybe (error "errorSpan: no position information") . findHint matchSpan
+  where
+  matchSpan (PositionedError ss) = Just ss
+  matchSpan _ = Nothing
+
+-- | Get the module name for an error
+errorModule :: ErrorMessage -> ModuleName
+errorModule = fromMaybe (error "errorModule: no position information") . findHint matchModule
```
Comment (garyb) [inline src/Language/PureScript/Errors.hs]: Is it safe to do this? I know we _should_ have module info included with every warning and error, but I'm not 100% confident that they all do right now.

**Comment by garyb** [comment]
Looks good, modulo comments :+1:

**Comment by nwolverson** [comment]
Good start...

The message still includes the lines "in module Blah" "at file.purs line...." which is now redundant. Also the _See wiki_ line which I don't think is required when the error code is provided.

File: src/Language/PureScript/Errors.hs (hunk 2 above)
Comment (paf31) [inline src/Language/PureScript/Errors.hs]: Probably not. We should really change the error message type to enforce these few things, but for now, I'll just change this to use `Maybe`.

**Comment by paf31** [comment]
@kRITZCREEK Yeah, I'll probably pull that out into its own `JSON` module.

@nwolverson I'll try to remove those.

Thanks everyone for the feedback on this.

**Comment by FrigoEU** [comment]
Hey, I'm super excited about this so I'm working with this branch now and made a syntastic (for vim) plugin for it. I noticed that there's some things not working yet. Should I log them here, or is it known that it's not finished yet? Maybe take a stab at it myself?

**Comment by paf31** [comment]
I just pushed a couple of fixes, but if you see anything else, yes, please log it.

**Comment by FrigoEU** [comment]
Just a few things I found:

1. "The type declaration for x should be followed by its definition": The line/col numbers in the JSON error point to the first statement after the module definition, while the line/col numbers in the text error point to the correct lines.
2. Parsing errors don't seem to return json errors yet.
3. As mentioned above, the file, line and column information in the message property is less useful in JSON errors.
4. Type unification errors don't have the correct position info in the json position property, while they do have the correct one in the error text.

**Comment by paf31** [comment]
@FrigoEU Thanks! I'm looking forward to syntastic support. The error locations should be fixed now.

**Comment by paf31** [comment]
I think this is ready for review again. I just need to fix the issue reported by @nwolverson.

Another idea was to use the JSON response to report which modules were rebuilt. I'm not sure if we should deal with that here or not.

**Comment by FrigoEU** [comment]
The issues I reported above seem to all be fixed.

**Comment by paf31** [comment]
@garyb Shall I merge this one then?

**Comment by garyb** [comment]
Yeah, definitely! :+1:

**Pull request merged**